injection/sharedmain/main.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sharedmain
+
+import (
+	"flag"
+	"log"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	// Uncomment the following line to load the gcp plugin (only required to authenticate against GKE clusters).
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+
+	"github.com/knative/pkg/configmap"
+	"github.com/knative/pkg/controller"
+	"github.com/knative/pkg/injection"
+	"github.com/knative/pkg/injection/clients/kubeclient"
+	"github.com/knative/pkg/logging"
+	"github.com/knative/pkg/metrics"
+	"github.com/knative/pkg/signals"
+	"github.com/knative/pkg/system"
+	"go.uber.org/zap"
+)
+
+const (
+	component = "controller"
+)
+
+var (
+	masterURL  = flag.String("master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	kubeconfig = flag.String("kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+)
+
+func Main() {
```

Comment:
Can we add some flexibility here wrt. to enabling customization.

For example - I may want to expose some runtime stats using https://golang.org/pkg/expvar - but I need the context (from signals) to help coordinate shutdown (or I pass one in) etc...

Reply:
Yeah, we can finesse this a bit to improve it's flexibility, this was just a starting point.

Reply:
I added a few variations to let folks customize aspects of this while sharing the meat of it.